File: frameworks/typescript-cypress/services/product_service.ts
import * as allure from 'allure-js-commons';
import { environment } from '../config/environment';
import { guardedApi, type ApiResponse } from '../helpers/api_response';
import { ACTIONS, step } from '../helpers/steps';

export type ProductInput = {
  name: string;
  description: string;
  price: number;
  stock: number;
  category: string;
};

export type Product = ProductInput & { id: number; permissions: string };

export const authHeaders = (): Record<string, string> => ({
  Authorization: `Bearer ${environment.apiKey()}`,
  'Content-Type': 'application/json',
});

export class ProductService {
  clearCart(): Cypress.Chainable<void> {
    return step(ACTIONS.apiClearCart, () => {
      guardedApi({ method: 'DELETE', url: '/api/cart', headers: authHeaders() }).then(
        (response) => {
          expect(response.status, 'clear cart status').to.eq(200);
        },
      );
    });
  }

  createProduct(input: ProductInput): Cypress.Chainable<Product> {
    return step(ACTIONS.apiCreateProduct, () =>
      guardedApi<{ data: Product }>({
        method: 'POST',
        url: '/api/products',
        headers: authHeaders(),
        body: input,
      }).then((response) => {
        this.attachExchange('Create product', input, response);
        expect(response.status, 'create product status').to.eq(201);
        return response.body.data;
      }),
    );
  }

  getProduct(id: number): Cypress.Chainable<Product> {
    return step(ACTIONS.apiGetProduct, () =>
      guardedApi<{ data: Product }>({
        url: `/api/products/${id}`,
        headers: authHeaders(),
      }).then((response) => {
        expect(response.status, 'read product status').to.eq(200);
        return response.body.data;
      }),
    );
  }

  updateProduct(id: number, input: Partial<ProductInput>): Cypress.Chainable<Product> {
    return step(ACTIONS.apiUpdateProduct, () =>
      guardedApi<{ data: Product }>({
        method: 'PATCH',
        url: `/api/products/${id}`,
        headers: authHeaders(),
        body: input,
      }).then((response) => {
        expect(response.status, 'update product status').to.eq(200);
        return response.body.data;
      }),
    );
  }

  deleteOrder(id: number): Cypress.Chainable<void> {
    return step(ACTIONS.apiDeleteOrder, () => {
      guardedApi({
        method: 'DELETE',
        url: `/api/orders/${id}`,
        headers: authHeaders(),
      }).then((response) => {
        expect([200, 404], 'delete order status').to.include(response.status);
      });
    });
  }

  deleteProduct(id: number): Cypress.Chainable<void> {
    return step(ACTIONS.apiDeleteProduct, () => {
      guardedApi({
        method: 'DELETE',
        url: `/api/products/${id}?force=true`,
        headers: authHeaders(),
      }).then((response) => {
        expect([200, 404], 'delete product status').to.include(response.status);
      });
    });
  }

  /** Reports the exchange without the credentials that authorized it. */
  private attachExchange(name: string, requestBody: unknown, response: ApiResponse): void {
    allure.attachment(
      name,
      JSON.stringify({ request: requestBody, response: { status: response.status } }, null, 2),
      'application/json',
    );
  }
}

File: frameowkrs/python-playwright/services/product_service.py
from typing import Any

import allure
import requests

from config.environment import environment


class ProductService:
    def __init__(self) -> None:
        self.session = requests.Session()
        self.session.headers.update(
            {"Authorization": f"Bearer {environment.api_key}", "Content-Type": "application/json"}
        )

    def clear_cart(self) -> None:
        self.session.delete(f"{environment.base_url}/api/cart", timeout=20)

    def create_product(self, product: dict[str, Any]) -> dict[str, Any]:
        response = self.session.post(
            f"{environment.base_url}/api/products", json=product, timeout=20
        )
        allure.attach(
            str({"request": product, "response_status": response.status_code}),
            "Create product",
            allure.attachment_type.JSON,
        )
        assert response.status_code == 201, response.text
        return response.json()["data"]

    def get_product(self, product_id: int) -> dict[str, Any]:
        response = self.session.get(
            f"{environment.base_url}/api/products/{product_id}", timeout=20
        )
        response.raise_for_status()
        return response.json()["data"]

    def update_product(self, product_id: int, changes: dict[str, Any]) -> dict[str, Any]:
        response = self.session.patch(
            f"{environment.base_url}/api/products/{product_id}", json=changes, timeout=20
        )
        response.raise_for_status()
        return response.json()["data"]

    def delete_order(self, order_id: int) -> None:
        response = self.session.delete(
            f"{environment.base_url}/api/orders/{order_id}", timeout=20
        )
        assert response.status_code in (200, 404), response.text

    def delete_product(self, product_id: int) -> None:
        response = self.session.delete(
            f"{environment.base_url}/api/products/{product_id}?force=true", timeout=20
        )
        assert response.status_code in (200, 404), response.text
